### PR TITLE
[Merged by Bors] - Add `Max Partition Size` property to topic

### LINKED ISSUE #2179: Storage: Enforce size based retention for topic
The current retention policy only supports the time-based property.   This may result in a incorrect estimate of the total data size.  We should add max data size properties as part of retention. 

The topic should have new property: 'max_size' to indicate maximum data size will be allowed to accumulate.

### PR DESCRIPTION
**Pull request by galibey**
This PR includes the following changes:
1. Added `max_partition_size` parameter to `fluvio-cli` and topic and partition specs.
2. `Cleaner` in `fluvio-storage` is responsible for enforcing replica size. Replica size is index size + log file size. If the size is exceeded so the first segment is removed.
3. `Cleaner` was moved from segments level to replica level to be able to access replica size.
4. Introduced `Size64` type as the partition size can easily overflow current `Size`.
5. `human_bytes` crate is replaced by `bytesize`. The latter is able to convert bytes to string and from string whereas the former only to string.

Closes #2179

### CODE REVIEW

**Comment by sehz** [comment]
`check crate-version` job in CI is failing

File: crates/fluvio-storage/src/cleaner.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        let replica_size = self.replica_size.get();
+        let max_partition_size = self.replica_config.max_partition_size.get();
+        let excess = replica_size.saturating_sub(max_partition_size);
+        if excess > 0 {
+            let segment_size = self.replica_config.segment_max_bytes.get() as u64;
+            let mut count_to_remove = excess.div(segment_size);
+            if excess.rem(segment_size) > 0 {
+                count_to_remove += 1;
+            }
+            info!("Replica size = {} bytes exceeded max partition size = {}. {} segments will be removed.",
```
Comment (sehz) [inline crates/fluvio-storage/src/cleaner.rs:95]: can you change to structure format?
```
info!(replica_size, segments, "....")
```

File: crates/fluvio-storage/src/mut_index.rs [hunk 2]
```diff
@@ -253,6 +258,12 @@ impl Index for MutLogIndex {
     }
 
     fn len(&self) -> Size {
+        self.first_empty_slot
+            .checked_mul(INDEX_ENTRY_SIZE)
+            .expect("index size must not exceed u32::MAX")
```
Comment (sehz) [inline crates/fluvio-storage/src/mut_index.rs:263]: we should not have expect in the non test code

Comment (sehz) [inline crates/fluvio-storage/src/mut_index.rs:263]: we should not have expect in the non test code

File: crates/fluvio-storage/src/mut_index.rs [hunk 3]
```diff
@@ -253,6 +258,12 @@ impl Index for MutLogIndex {
...
+    }
+
+    fn cap(&self) -> Size {
```
Comment (sehz) [inline crates/fluvio-storage/src/mut_index.rs:266]: Can you make it more descriptive?

Comment (sehz) [inline crates/fluvio-storage/src/mut_index.rs:266]: Can you make it more descriptive?

**Review by sehz — CHANGES_REQUESTED** [review]
Looks very good.  Couple of minor nits on first pass.

File: crates/fluvio-dataplane-protocol/src/common.rs [hunk 4]
```diff
@@ -5,6 +5,7 @@ use crate::derive::{Encoder, Decoder};
 
 pub type Offset = i64;
 pub type Size = u32;
+pub type Size64 = u64;
```
Comment (sehz) [inline crates/fluvio-dataplane-protocol/src/common.rs:9]: need to bump up crate version

File: crates/fluvio-protocol/src/core/encoder.rs [hunk 5]
```diff
@@ -460,6 +480,71 @@ mod test {
...
+        let result = value.encode(&mut dest, 0);
+        assert!(result.is_ok());
+        assert_eq!(
+            dest,
+            vec![0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10]
+        );
+        assert_eq!(value.write_size(0), 9);
+    }
+
     #[test]
```
Comment (sehz) [inline crates/fluvio-protocol/src/core/encoder.rs:548]: bump up crate version

File: crates/fluvio-types/src/defaults.rs [hunk 6]
```diff
@@ -33,7 +33,11 @@ pub const SPU_LOG_INDEX_MAX_BYTES: u32 = 10485760;
 pub const SPU_LOG_INDEX_MAX_INTERVAL_BYTES: u32 = 4096;
 pub const SPU_LOG_SEGMENT_MAX_BYTES: u32 = 1073741824;
 
+pub const SPU_PARTITION_MAX_BYTES: u64 = 107_374_182_400; //100Gb
+pub const SPU_PARTITION_MAX_BYTES_MIN: u64 = SPU_LOG_LOG_SEGMENT_MAX_BYTE_MIN as u64 * 2;
+
```
Comment (sehz) [inline crates/fluvio-types/src/defaults.rs:38]: bump up crate version

File: k8-util/helm/fluvio-sys/templates/crd_topic.yaml [hunk 7]
```diff
@@ -170,6 +172,14 @@ spec:
             type: string
             description: Topic Status
             jsonPath: .status.resolution
+          - name: Segment Size
+            type: integer
+            description: Segment Size
+            jsonPath: .spec.storage.segmentSize
+          - name: Max Partition Size
+            type: integer
+            description: Max Partition Size
```
Comment (sehz) [inline k8-util/helm/fluvio-sys/templates/crd_topic.yaml:181]: bump up chart version

**Review by sehz — CHANGES_REQUESTED** [review]
2nd pass

File: crates/fluvio-storage/src/cleaner.rs (hunk 1 above)
Comment (galibey) [reply crates/fluvio-storage/src/cleaner.rs]: Done

File: crates/fluvio-storage/src/mut_index.rs (hunk 2 above)
Comment (galibey) [reply crates/fluvio-storage/src/mut_index.rs]: Done

Comment (galibey) [reply crates/fluvio-storage/src/mut_index.rs]: Done

File: crates/fluvio-storage/src/mut_index.rs (hunk 3 above)
Comment (galibey) [reply crates/fluvio-storage/src/mut_index.rs]: Done

Comment (galibey) [reply crates/fluvio-storage/src/mut_index.rs]: Done

File: crates/fluvio-storage/src/cleaner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (galibey) [reply crates/fluvio-storage/src/cleaner.rs]: Done

File: crates/fluvio-dataplane-protocol/src/common.rs (hunk 4 above)
Comment (galibey) [reply crates/fluvio-dataplane-protocol/src/common.rs:9]: Done

File: crates/fluvio-protocol/src/core/encoder.rs (hunk 5 above)
Comment (galibey) [reply crates/fluvio-protocol/src/core/encoder.rs:548]: Done

File: crates/fluvio-types/src/defaults.rs (hunk 6 above)
Comment (galibey) [reply crates/fluvio-types/src/defaults.rs:38]: Done

File: k8-util/helm/fluvio-sys/templates/crd_topic.yaml (hunk 7 above)
Comment (galibey) [reply k8-util/helm/fluvio-sys/templates/crd_topic.yaml:181]: Done

**Review by sehz — APPROVED** [review]
Nice work.  LGTM

**Pull request closed without merging**